src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -627,16 +629,18 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
                 if (const auto * rhs_literal = equals_arguments[1]->as<ConstantNode>())
                 {
                     args.push_back(rhs_literal->getValue());
+                    tuple_element_types.push_back(rhs_literal->getResultType());
                 }
                 else
                 {
                     const auto * lhs_literal = equals_arguments[0]->as<ConstantNode>();
                     assert(lhs_literal);
                     args.push_back(lhs_literal->getValue());
+                    tuple_element_types.push_back(lhs_literal->getResultType());
                 }
             }
 
-            auto rhs_node = std::make_shared<ConstantNode>(std::move(args));
+            auto rhs_node = std::make_shared<ConstantNode>(std::move(args), std::make_shared<DataTypeTuple>(std::move(tuple_element_types)));
```

Comment:
Doesn't it also require adding a conversation of tuple elements to IN first argument type? I'm not sure if the semantics is correct right now.

Reply:
I'm not quite sure what this question is about, but regarding this fix - the issue is that `DB::Field` doesn't have `DateTime` types (replacing them by some Decimal I believe), so `ConstantNode` created without explicit type argument will derive types from `Field`'s in `args`